### ISSUE TITLE
[clj] Implement conj

### ISSUE THREAD

**Issue by yurrriq** [lfe/lfe#268]
> Treat vectors as tuples.

https://github.com/lfex/clj/issues/17

**Comment by rvirding** [comment]
I don't know, I am not really taking in developing the clj module.

**Comment by yurrriq** [comment]
Yeah, I think the idea is to behave similarly to `clj:get-in/2` and treat tuples as Clojure does vectors.

**Comment by yurrriq** [comment]
I had a partial implementation floating around at one point, but I'm not sure if anything useful ever came from it.

**Comment by skovsgaard** [comment]
I get it and absolutely agree on treating tuples as vectors, but what I was wondering was that since Clojure `conj` behaves like this with maps:

``` clojure
=>(conj {:a 1 :b 2} {:c 3})
{:a 1, :b 2, :c 3}
```

should we try to do something similar with Erlang maps so the following is valid use of `clj:conj`?

``` lfe
>(clj:conj '#m(a 1 b 2) '#m(c 3))
#M(c 3 b 2 a 1)
```

Exactly what `maps:merge/2` does now but built in as a sort-of-polymorphism in `conj`?

**Comment by yurrriq** [comment]
The examples are double quoted, but yeah, that looks good to me.

We should support many key/values too:

``` clj
=> (conj {:a 1, :b 2} {:c 3, :d 4})
{:b 2, :a 1, :c 3, :d 4}
```

``` lfe
> (maps:merge #m(a 1 b 2) #m(c 3 d 4))
#M(d 4 c 3 b 2 a 1)
```

**Comment by skovsgaard** [comment]
Yup, basically so that `when is_map`, `clj:conj/2` behaves like `maps:merge/2`?

**Comment by yurrriq** [comment]
Sounds good to me.

**Comment by pigmej** [comment]
```
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          From: Eric BaileySent: poniedziałek, 24 października 2016 21:17To: rvirding/lfeReply To: rvirding/lfeSubject: Re: [rvirding/lfe] [clj] Implement conj (#268)Yeah, I think the idea is to behave similarly to clj:get-in/2 and treat tuples as Clojure does vectors.
```

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.

{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/rvirding/lfe","title":"rvirding/lfe","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/rvirding/lfe"}},"updates":{"snippets":[{"icon":"PERSON","message":"@yurrriq in #268: Yeah, I think the idea is to behave similarly to `clj:get-in/2` and treat tuples as Clojure does vectors."}],"action":{"name":"View Issue","url":"https://github.com/rvirding/lfe/issues/268#issuecomment-255837930"}}}‎

**Comment by yurrriq** [comment]
@pigmej, looks like some kind of third-party app error.

**Comment by skovsgaard** [comment]
Here's what I have so far. It works like Clojure's `conj` with the exception that I can't work out how to make it variadic for maps the way it is for lists. Any and all suggestions are welcome.

``` lisp
(defmacro conj
  (`[,coll . ,xs]
   `(cond ((is_list ,coll) (lists:flatten (cons (list ,@xs) ,coll)))
          ((is_tuple ,coll) (lists:flatten (cons (list ,@xs)
                                                 (tuple_to_list ,coll))))
          ((is_map ,coll) (maps:merge (car (list ,@xs)) ,coll)))))
```

**Comment by yurrriq** [comment]
Maybe some sort of fold would help with the `maps` clause. We want to support R16 still, I think (/cc @rvirding), so you may need to jump through a few extra hoops. I think the private implementation of `clj:get-in/2` has a good example.

**Comment by skovsgaard** [comment]
I submitted #294 which implements it as variadic for both lists, maps, and tuples. The latter of which appends values the same way it would for vectors in Clojure.

Update: Of course R16 fails. I'll leave the PR open while I fix that somehow.

**Comment by yurrriq** [comment]
This can be closed now that #294 is merged.
